### PR TITLE
Raise AttributeError when missing Schema methods

### PR DESCRIPTION
**Pull request by ghost**
Currently mistyping a schema method name when using `fields.Method` will result in a `ValueError` rather than throwing an `AttributeError` that details what went wrong:

```python
class MySchema(Schema):
    # Method constructor argument is misspelled but actual method is not
    foobar = fields.Method("get_footbar")

    def foobar(self, obj):
         return obj.some_other_data()
```

```
  File ".../lib/python3.8/site-packages/marshmallow/schema.py", line 557, in dump
    result = self._serialize(processed_obj, many=many)
  File ".../lib/python3.8/site-packages/marshmallow/schema.py", line 521, in _serialize
    value = field_obj.serialize(attr_name, obj, accessor=self.get_attribute)
  File ".../lib/python3.8/site-packages/marshmallow/fields.py", line 312, in serialize
    return self._serialize(value, attr, obj, **kwargs)
  File ".../lib/python3.8/site-packages/marshmallow/fields.py", line 1663, in _serialize
    method = utils.callable_or_raise(
  File ".../lib/python3.8/site-packages/marshmallow/utils.py", line 285, in callable_or_raise
    raise ValueError("Object {!r} is not callable.".format(obj))
ValueError: Object None is not callable.
```

With this change the error would yield something like:

```python
AttributeError: 'MySchema' object has no attribute 'get_footbar'
```

https://github.com/marshmallow-code/marshmallow/issues/1674

### CODE REVIEW

**Comment by lafrech** [comment]
Shouldn't the `callable_or_raise` check be performed at init time?

- Better for performance
- Better for debug, no runtime error

Or do we expect users to patch their schema and add the method after init?

**Comment by ghost** [comment]
> Shouldn't the `callable_or_raise` check be performed at init time?
> 
> * Better for performance
> * Better for debug, no runtime error
> 
> Or do we expect users to patch their schema and add the method after init?

@lafrech I just pushed a commit 5969cf4 that adds a `Method._bind_to_schema()` override and performs the callable checks there rather than having them be performed at serialization/deserialization time. Let me know your thoughts!

Currently `callable_or_raise()` will raise `ValueError` if the attribute type is incorrect (not callable). Shouldn't this be a [`TypeError` based on the use case?](https://docs.python.org/3/library/exceptions.html#TypeError)

I don't think patching the schema afterwards should be expected behavior, but if they were to swap the attribute after init and make it non-callable they would receive `TypeError` in that situation rather than `ValueError` (with the changes of 5969cf4).

**Comment by ghost** [comment]
Hi @lafrech 

I've just updated `callable_or_raise()` in c2764b7 to raise `TypeError` rather than `ValueError` as to be consistent with python when objects are not callable:

```python
>>> 1()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not callable
```

File: src/marshmallow/utils.py [hunk 1]
```diff
@@ -280,7 +280,7 @@ def set_value(dct: typing.Dict[str, typing.Any], key: str, value: typing.Any):
 def callable_or_raise(obj):
     """Check that an object is callable, else raise a :exc:`ValueError`."""
```
Comment (lafrech) [inline src/marshmallow/utils.py]: s/ValueError/TypeError/

File: tests/test_serialization.py [hunk 2]
```diff
@@ -356,7 +356,7 @@ def test_boolean_field_serialization(self, user):
         assert field.serialize("none", user) is None
 
     def test_function_with_uncallable_param(self):
-        with pytest.raises(ValueError):
+        with pytest.raises(TypeError):
             fields.Function("uncallable")
```
Comment (lafrech) [inline tests/test_serialization.py]: Isn't this test a duplicate of `test_function_field_passed_uncallable_object`?

File: tests/test_serialization.py [hunk 3]
```diff
@@ -412,7 +412,7 @@ class BadSerializer(Schema):
             bad_field = fields.Method("invalid")
 
         u = User("Foo")
-        with pytest.raises(ValueError):
+        with pytest.raises(AttributeError):
             BadSerializer().dump(u)
```
Comment (lafrech) [inline tests/test_serialization.py]: Is the call to `dump()` necessary for the error to be raised?

File: tests/test_serialization.py [hunk 4]
```diff
@@ -431,7 +431,7 @@ class BadSerializer(Schema):
             bad_field = fields.Method("foo")
 
         u = User("Foo")
-        with pytest.raises(ValueError):
+        with pytest.raises(TypeError):
             BadSerializer().dump(u)
```
Comment (lafrech) [inline tests/test_serialization.py]: Is the call to `dump()` necessary for the error to be raised?

**Review by lafrech — COMMENTED** [review]
Thanks.

Looks good to me, apart from a few comments.

I don't think it qualifies as a breaking change so it could be introduced in the next 3.x.

I don't mind a second reviewer on this.

File: tests/test_serialization.py (hunk 2 above)
Comment (ghost) [reply tests/test_serialization.py]: Yeah looks like it, I'll go ahead and remove this one

File: tests/test_serialization.py (hunk 3 above)
Comment (ghost) [reply tests/test_serialization.py]: Good catch, no `dump()` will not be necessary anymore

**Comment by ghost** [comment]
@lafrech thank you! Updated in 124fad7 to address the comments

**Comment by lafrech** [comment]
While we're at it, wouldn't it make sense to get the methods in `_bind_to_schema` and cache them in attributes?

```py
    def __init__(self, serialize: str = None, deserialize: str = None, **kwargs):
        # Set dump_only and load_only based on arguments
        kwargs["dump_only"] = bool(serialize) and not bool(deserialize)
        kwargs["load_only"] = bool(deserialize) and not bool(serialize)
        super().__init__(**kwargs)
        self.serialize_method_name = serialize
        self.deserialize_method_name = deserialize
        self._serialize_method = None
        self._deserialize_method = None

    def _bind_to_schema(self, field_name, schema):
        if self.serialize_method_name:
            self._serialize_method = utils.callable_or_raise(getattr(schema, self.serialize_method_name))

        if self.deserialize_method_name:
            self._deserialize_method = utils.callable_or_raise(getattr(schema, self.deserialize_method_name))

        super()._bind_to_schema(field_name, schema)

    def _serialize(self, value, attr, obj, **kwargs):
        if not self._serialize_method:
            return missing_
        return self._serialize_method(obj)

    def _deserialize(self, value, attr, data, **kwargs):
        if self._deserialize_method:
            return self.deserialize_method(value)
        return value
```

**Comment by ghost** [comment]
> While we're at it, wouldn't it make sense to get the methods in `_bind_to_schema` and cache them in attributes?

Good call, yeah that's a much better approach and also addresses the attribute availability thing I mentioned before: users shouldn't monkey-patch schema methods. I updated this in b48e7da

There's one issue I encountered with that though is there is this test:

```python
def test_method_field_deserialize_only(self):
        class MethodDeserializeOnly(Schema):
            def lowercase_name(self, value):
                return value.lower()

        m = fields.Method(deserialize="lowercase_name")
        m.parent = MethodDeserializeOnly()

        assert m.deserialize("ALEC") == "alec"
```

From the way this test is written, `m` never has `_bind_to_schema()` called because it's a field completely outside the declaration of the schema itself... is this something that's supported by Marshmallow? 

I'm a bit confused on the use case, but I rewrote the test to utilize the Schema's `load()` instead:

```python
def test_method_field_deserialize_only(self):
        class MethodDeserializeOnly(Schema):
            name = fields.Method(deserialize="lowercase_name")

            def lowercase_name(self, value):
                return value.lower()

        assert MethodDeserializeOnly().load({"name": "ALEC"})["name"] == "alec"
```

Let me know your thoughts

**Comment by lafrech** [comment]
This test already relied on

```py
    m.parent = MethodDeserializeOnly()
```

which is basically what `_bind_to_schema` does.

Apart from this PR, it would be a good idea to rewrite the test to use `_bind_to_schema` explicitly IMHO.

There's no way this field can be independent of the schema.

I'm fine with this change.

Note that before your last commit, the use case was broken already because the field didn't go through the callable or raise test, and this wasn't detected due to the test being badly written. Another reason to rewrite it.

Is the independent field use case supported? Well, I've been looking through the fields module, and the other fields actually using bind_to_schema are `DateTime` to get a format from the field class, and container/nested fields to pass `_bind_to_schema`. `DateTime` field doesn't need it to work.

So I suppose it works and could be supported in the general case but it makes no sense, by definition, for the `Method` field so I wouldn't mind making it an exception if the rule was to support the use case.

Regarding your implementation with `hasattr` vs. my proposal, unless there is a reason I'm missing, I tend to apply the general rule of declaring all attributes in `__init__`. pylint, for instance, complains about setting attributes in other methods as it is considered bad practice. Totally opinion-based. My proposal should be rewritten `if self._serialize_method is None`, though, to be more explicit and to skip a boolean evaluation.

**Comment by ghost** [comment]
@lafrech awesome, thanks for the explanation!

I force pushed the last commit to 6d0edf8 to swap the `hasattr()` for the `None` init + `is not None` checks, I agree it's much better to explicitly create the attribute if it's expected to be used

**Review by lafrech — APPROVED** [review]
Great. Thanks for your work on this.

One last thing. Please also add yourself to AUTHORS file and this should be good to go.

Still, requesting review from @sloria.

**Comment by ghost** [comment]
@sloria forgot that `Authors.rst` got updated recently, so just fixed the conflict. Should be good to review now!

**Review by sloria — APPROVED** [review]

**Pull request merged**